torch/csrc/cuda/nccl.cpp

```diff
@@ -121,232 +129,330 @@ static void _check_inputs(std::vector<at::Tensor> &inputs, std::vector<at::Tenso
-  case at::TypeID::CUDALong    : return ncclInt64;
-  case at::TypeID::CUDAInt     : return ncclInt;
-  case at::TypeID::CUDAChar    : return ncclChar;
-  case at::TypeID::CUDAByte    : return ncclChar;
+static ncclDataType_t _get_data_type(const Type& type) {
+  if (type.backend() != kCUDA) {
+    throw std::runtime_error("Unconvertible NCCL type");
+  }
+  switch (type.scalarType()) {
+  case at::kFloat   : return ncclFloat;
+  case at::kHalf    : return ncclHalf;
+  case at::kDouble  : return ncclDouble;
+  case at::kLong    : return ncclInt64;
+  case at::kInt     : return ncclInt;
+  case at::kChar    : return ncclChar;
+  case at::kByte    : return ncclChar;
   default: throw std::runtime_error("Unconvertible NCCL type");
   }
 }
 
+static void _start_group() {
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  CHECK(ncclGroupStart());
+#endif
+}
+
+static void _end_group() {
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR >= 2)
+  CHECK(ncclGroupEnd());
+#endif
+}
+
+PyObject * THCPModule_nccl_version(PyObject *self, PyObject *args) {
+#if defined(NCCL_MAJOR)
+  return PyInt_FromLong(NCCL_MAJOR * 1000 + NCCL_MINOR * 100 + NCCL_PATCH);
+#else
+  return PyInt_FromLong(1000);  // assume NCCL 1.0
+#endif
+}
+
+PyObject * THCPModule_nccl_unique_id(PyObject *self, PyObject *args) {
+  HANDLE_TH_ERRORS
+  ncclUniqueId id;
+  CHECK(ncclGetUniqueId(&id));
+  return PyBytes_FromStringAndSize((char*)&id, NCCL_UNIQUE_ID_BYTES);
+  END_HANDLE_TH_ERRORS
+}
+
+static ncclComm_t unpack_nccl_comm(PyObject* capsule) {
+  ncclComm_t comm = (ncclComm_t)PyCapsule_GetPointer(capsule, COMM_CAPSULE_NAME);
+  if (!comm) throw python_error();
+  return comm;
+}
+
+static void destroy_nccl_comm(PyObject* capsule) {
+  HANDLE_TH_ERRORS
+  ncclComm_t comm = unpack_nccl_comm(capsule);
+  with_no_gil([&]{
+    ncclCommDestroy(comm);
+  });
+  END_HANDLE_TH_ERRORS_RET()
+}
+
+static std::vector<THCStream*> unpack_streams(PyObject* obj, size_t size) {
+  if (obj == Py_None) {
+    return std::vector<THCStream*>(size, nullptr);
+  }
+  auto streams = THPUtils_PySequence_to_THCStreamList(obj);
+  if (streams.size() != size) {
+    throw std::runtime_error("number of streams is not equal to number of inputs");
+  }
+  return streams;
+}
+
+static std::vector<ncclComm_t> unpack_comms(PyObject* obj, size_t size) {
+  if (obj == Py_None) {
+    return std::vector<ncclComm_t>();
+  }
+  std::vector<ncclComm_t> comms;
+  if (PyCapsule_CheckExact(obj)) {
+    comms = { unpack_nccl_comm(obj) };
+  } else {
+    auto seq = THPObjectPtr(PySequence_Fast(obj, "comm is not a sequence"));
+    if (!seq) throw python_error();
+    auto size = PySequence_Fast_GET_SIZE(seq.get());
+    comms = std::vector<ncclComm_t>(size);
+    for (int64_t i = 0; i < size; i++) {
+      comms[i] = unpack_nccl_comm(PySequence_Fast_GET_ITEM(seq.get(), i));
+    }
+  }
+  if (comms.size() != size) {
+    throw std::runtime_error("number of communicators is not equal to number of inputs");
```

Comment:
Uh now I see that the condition in `_get_communicators` will always be satisfied, but I still think that doing it like this:

```cpp
auto comms = user_comms.empty() ? _get_communicator(inputs) : user_comms;
```

would be much clearer.